File: services/agents/lib/key_health.py
"""Global fallback API key exhaustion notifier.

Single chokepoint called by provider clients (openrouter, tavily,
semantic_scholar, chandra) when a request using a *fallback* env-var key
returns 401, 402, or sustained 429. UPSERTs into `provider_key_alerts`
(shared with apps/km TS notifier) and emails the operator via Resend when
the per-row dedup window has elapsed.

BYOK (per-user) key failures must NOT call this — only env-fallback paths.
"""
from __future__ import annotations

import logging
import os
from datetime import datetime, timedelta, timezone
from typing import Any, Literal

import httpx

logger = logging.getLogger(__name__)

Reason = Literal["key_invalid", "key_exhausted", "key_rate_limited"]

DEDUP_WINDOW = timedelta(hours=1)
RATE_LIMIT_THRESHOLD_HITS = 5
RATE_LIMIT_WINDOW = timedelta(minutes=10)

_QUOTA_HINTS = (
    "insufficient_quota",
    "insufficient credit",
    "insufficient credits",
    "payment_required",
    "out of credit",
    "balance",
    "quota exceeded",
    "credit limit",
    "key limit",
    "more credits",
    "fewer max_tokens",
)


def classify_provider_error(status_code: int, body_text: str) -> Reason | None:
    """Map an HTTP status + response body to an alert reason, or None.

    GSD-136: OR returns HTTP 401 (NOT 402) when a Provisioning-API key has
    its `limit` exhausted. The body carries a "more credits" / "credit
    limit" quota hint. We disambiguate by body content: hint present →
    key_exhausted, hint absent → key_invalid (real auth failure).
    """
    body = (body_text or "").lower()
    has_quota_hint = any(h in body for h in _QUOTA_HINTS)
    if status_code == 401:
        return "key_exhausted" if has_quota_hint else "key_invalid"
    if status_code == 402:
        return "key_exhausted"
    if status_code == 403 and has_quota_hint:
        return "key_exhausted"
    if status_code == 429:
        return "key_rate_limited"
    return None


_UPSERT_SQL = """
INSERT INTO provider_key_alerts (provider, env_var, reason, hit_count, sample_error)
VALUES ($1, $2, $3, 1, $4)
ON CONFLICT (provider, env_var, reason) WHERE cleared_at IS NULL
DO UPDATE SET
  hit_count = CASE
    WHEN $3 = 'key_rate_limited'
         AND provider_key_alerts.first_seen_at < NOW() - INTERVAL '10 minutes'
    THEN 1
    ELSE provider_key_alerts.hit_count + 1
  END,
  first_seen_at = CASE
    WHEN $3 = 'key_rate_limited'
         AND provider_key_alerts.first_seen_at < NOW() - INTERVAL '10 minutes'
    THEN NOW()
    ELSE provider_key_alerts.first_seen_at
  END,
  last_seen_at = NOW(),
  sample_error = COALESCE(EXCLUDED.sample_error, provider_key_alerts.sample_error)
RETURNING id, hit_count, first_seen_at, last_seen_at, last_alerted_at
"""

_MARK_ALERTED_SQL = (
    "UPDATE provider_key_alerts SET last_alerted_at = NOW() WHERE id = $1"
)


def _should_alert(reason: Reason, row: dict[str, Any]) -> bool:
    last_alerted = row.get("last_alerted_at")
    now = datetime.now(timezone.utc)
    if last_alerted is not None and now - last_alerted < DEDUP_WINDOW:
        return False
    if reason == "key_rate_limited":
        first_seen = row.get("first_seen_at")
        if first_seen is None or now - first_seen > RATE_LIMIT_WINDOW:
            return False
        if row.get("hit_count", 0) < RATE_LIMIT_THRESHOLD_HITS:
            return False
    return True


async def _send_resend_email(
    *,
    provider: str,
    env_var: str,
    reason: Reason,
    row: dict[str, Any],
    sample_error: str | None,
) -> bool:
    api_key = os.environ.get("RESEND_API_KEY", "").strip()
    if not api_key:
        logger.warning(
            "provider key alert suppressed — RESEND_API_KEY unset "
            "(provider=%s env=%s reason=%s)",
            provider,
            env_var,
            reason,
        )
        return False
    to_addr = os.environ.get("ALERT_EMAIL_TO", "mohidfbutt@gmail.com").strip()
    from_addr = os.environ.get(
        "ALERT_EMAIL_FROM", "alerts@tryepisteme.com"
    ).strip()

    subject = f"[episteme] {provider} key {reason} — {env_var}"
    body_lines = [
        f"Provider: {provider}",
        f"Env var: {env_var}",
        f"Reason: {reason}",
        f"Hit count (since first_seen): {row.get('hit_count')}",
        f"First seen: {row.get('first_seen_at')}",
        f"Last seen: {row.get('last_seen_at')}",
        f"Last alerted: {row.get('last_alerted_at')}",
        "",
        "Sample error:",
        (sample_error or "").strip()[:1000],
    ]
    payload = {
        "from": from_addr,
        "to": [to_addr],
        "subject": subject,
        "text": "\n".join(body_lines),
    }
    try:
        async with httpx.AsyncClient(timeout=10) as c:
            r = await c.post(
                "https://api.resend.com/emails",
                headers={
                    "Authorization": f"Bearer {api_key}",
                    "Content-Type": "application/json",
                },
                json=payload,
            )
            r.raise_for_status()
    except Exception:
        logger.exception(
            "resend POST failed (provider=%s env=%s)", provider, env_var
        )
        return False
    return True


async def record_and_maybe_alert(
    pool: Any,
    *,
    provider: str,
    env_var: str,
    reason: Reason,
    sample_error: str | None = None,
) -> bool:
    """UPSERT alert row + email operator if dedup window allows.

    Returns True iff a Resend email was successfully sent on this call.
    Safe to call from any provider error handler — never raises.
    """
    if pool is None:
        return False
    try:
        async with pool.acquire() as conn:
            row = await conn.fetchrow(
                _UPSERT_SQL, provider, env_var, reason, sample_error
            )
            if row is None:
                return False
            row_dict = dict(row)
            if not _should_alert(reason, row_dict):
                return False
            sent = await _send_resend_email(
                provider=provider,
                env_var=env_var,
                reason=reason,
                row=row_dict,
                sample_error=sample_error,
            )
            if sent:
                await conn.execute(_MARK_ALERTED_SQL, row_dict["id"])
            return sent
    except Exception:
        logger.exception(
            "record_and_maybe_alert failed (provider=%s env=%s reason=%s)",
            provider,
            env_var,
            reason,
        )
        return False

File: apps/km/src/lib/key-health.ts
import { sql } from "drizzle-orm";
import { db } from "@/lib/db";
import { sendEmail } from "@/lib/send-email";

/**
 * Post-fetch hook for KM routes that call OpenRouter directly with a key
 * resolved via BYOK-or-env. Fires the notifier ONLY when the failing key
 * came from the named env var (i.e. server fallback, not per-user BYOK).
 * Fire-and-forget: never throws, returns void.
 */
export function checkOpenRouterFallbackResponse(opts: {
  envVar: "EPISTEME_SHARED_LLM_KEY" | "OPENROUTER_API_KEY";
  apiKey: string;
  response: Response;
}): void {
  if (opts.response.ok) return;
  const envValue = process.env[opts.envVar];
  if (!envValue || opts.apiKey !== envValue) return;
  const reason = classifyProviderError(opts.response.status, "");
  if (!reason) {
    // Status alone wasn't enough — peek the body for quota hints (403 case).
    void opts.response
      .clone()
      .text()
      .then((text) => {
        const r2 = classifyProviderError(opts.response.status, text);
        if (r2) {
          void recordAndMaybeAlert({
            provider: "openrouter",
            envVar: opts.envVar,
            reason: r2,
            sampleError: text.slice(0, 1000),
          });
        }
      })
      .catch(() => {});
    return;
  }
  void opts.response
    .clone()
    .text()
    .then((text) => {
      void recordAndMaybeAlert({
        provider: "openrouter",
        envVar: opts.envVar,
        reason,
        sampleError: text.slice(0, 1000),
      });
    })
    .catch(() => {});
}

export type ProviderName =
  | "openrouter"
  | "tavily"
  | "semantic_scholar"
  | "chandra";

export type AlertReason =
  | "key_invalid"
  | "key_exhausted"
  | "key_rate_limited";

const DEDUP_WINDOW_MS = 60 * 60 * 1000;
const RATE_LIMIT_WINDOW_MS = 10 * 60 * 1000;
const RATE_LIMIT_THRESHOLD_HITS = 5;

const QUOTA_HINTS = [
  "insufficient_quota",
  "insufficient credit",
  "insufficient credits",
  "payment_required",
  "out of credit",
  "balance",
  "quota exceeded",
];

export function classifyProviderError(
  status: number,
  bodyText: string,
): AlertReason | null {
  const body = (bodyText ?? "").toLowerCase();
  if (status === 401) return "key_invalid";
  if (status === 402) return "key_exhausted";
  if (status === 403 && QUOTA_HINTS.some((h) => body.includes(h))) {
    return "key_exhausted";
  }
  if (status === 429) return "key_rate_limited";
  return null;
}

type AlertRow = {
  id: string;
  hit_count: number;
  first_seen_at: Date;
  last_seen_at: Date;
  last_alerted_at: Date | null;
};

function shouldAlert(reason: AlertReason, row: AlertRow): boolean {
  const now = Date.now();
  if (
    row.last_alerted_at &&
    now - new Date(row.last_alerted_at).getTime() < DEDUP_WINDOW_MS
  ) {
    return false;
  }
  if (reason === "key_rate_limited") {
    const firstSeenMs = new Date(row.first_seen_at).getTime();
    if (now - firstSeenMs > RATE_LIMIT_WINDOW_MS) return false;
    if (row.hit_count < RATE_LIMIT_THRESHOLD_HITS) return false;
  }
  return true;
}

async function sendResendEmail(opts: {
  provider: string;
  envVar: string;
  reason: AlertReason;
  row: AlertRow;
  sampleError: string | null;
}): Promise<boolean> {
  const to = (process.env.ALERT_EMAIL_TO ?? "mohidfbutt@gmail.com").trim();
  const subject = `[episteme] ${opts.provider} key ${opts.reason} — ${opts.envVar}`;
  const text = [
    `Provider: ${opts.provider}`,
    `Env var: ${opts.envVar}`,
    `Reason: ${opts.reason}`,
    `Hit count (since first_seen): ${opts.row.hit_count}`,
    `First seen: ${opts.row.first_seen_at}`,
    `Last seen: ${opts.row.last_seen_at}`,
    `Last alerted: ${opts.row.last_alerted_at}`,
    "",
    "Sample error:",
    (opts.sampleError ?? "").trim().slice(0, 1000),
  ].join("\n");

  return sendEmail({ to, subject, text });
}

export async function recordAndMaybeAlert(opts: {
  provider: ProviderName;
  envVar: string;
  reason: AlertReason;
  sampleError?: string | null;
}): Promise<boolean> {
  const sampleError = opts.sampleError ?? null;
  try {
    const upsert = await db.execute(sql`
      INSERT INTO provider_key_alerts (provider, env_var, reason, hit_count, sample_error)
      VALUES (${opts.provider}, ${opts.envVar}, ${opts.reason}, 1, ${sampleError})
      ON CONFLICT (provider, env_var, reason) WHERE cleared_at IS NULL
      DO UPDATE SET
        hit_count = CASE
          WHEN ${opts.reason} = 'key_rate_limited'
               AND provider_key_alerts.first_seen_at < NOW() - INTERVAL '10 minutes'
          THEN 1
          ELSE provider_key_alerts.hit_count + 1
        END,
        first_seen_at = CASE
          WHEN ${opts.reason} = 'key_rate_limited'
               AND provider_key_alerts.first_seen_at < NOW() - INTERVAL '10 minutes'
          THEN NOW()
          ELSE provider_key_alerts.first_seen_at
        END,
        last_seen_at = NOW(),
        sample_error = COALESCE(EXCLUDED.sample_error, provider_key_alerts.sample_error)
      RETURNING id, hit_count, first_seen_at, last_seen_at, last_alerted_at
    `);
    const rows = (upsert as unknown as AlertRow[]) ?? [];
    const row = rows[0];
    if (!row) return false;
    if (!shouldAlert(opts.reason, row)) return false;
    const sent = await sendResendEmail({
      provider: opts.provider,
      envVar: opts.envVar,
      reason: opts.reason,
      row,
      sampleError,
    });
    if (sent) {
      await db.execute(
        sql`UPDATE provider_key_alerts SET last_alerted_at = NOW() WHERE id = ${row.id}`,
      );
    }
    return sent;
  } catch (err) {
    console.error(
      "[key-health] recordAndMaybeAlert failed",
      { provider: opts.provider, envVar: opts.envVar, reason: opts.reason, err },
    );
    return false;
  }
}
